content/en/blogs/faas-openfunction.md

```diff
@@ -0,0 +1,300 @@
+
+### Function framework
+
+To make it easier to learn function specifications during development, we need to add a mechanism to transform function code to ready-to-run applications. This mechanism needs to be implemented by making a general-purpose main function that handles incoming requests through the `serving url` function.
+
+The main function contains a number of steps: one for associating user-submitted code, and the rest for common functionalities (such as context processing, event source handling, exception handling, port processing, etc.).
+
+During the process of building a function, the builder uses the main function template to render user code, and then builds the main function in the container image.
+
+Let's look at an example. Assuming we have the following function code:
+
+```go
+package hello
+
+import (
+    "fmt"
+    "net/http"
+)
+
+func HelloWorld(w http.ResponseWriter, r *http.Request) {
+    fmt.Fprint(w, "Hello, World!\n")
+}
+```
+
+After being converted by the function framework, the following application code will be generated:
+
+```go
+package main
+
+import (
+    "context"
+    "errors"
+    "fmt"
+    "github.com/OpenFunction/functions-framework-go/functionframeworks"
+    ofctx "github.com/OpenFunction/functions-framework-go/openfunction-context"
+    cloudevents "github.com/cloudevents/sdk-go/v2"
+    "log"
+    "main.go/userfunction"
+    "net/http"
+)
+
+func register(fn interface{}) error {
+    ctx := context.Background()
+    if fnHTTP, ok := fn.(func(http.ResponseWriter, *http.Request)); ok {
+        if err := functionframeworks.RegisterHTTPFunction(ctx, fnHTTP); err != nil {
+            return fmt.Errorf("Function failed to register: %v\n", err)
+        }
+    } else if fnCloudEvent, ok := fn.(func(context.Context, cloudevents.Event) error); ok {
+        if err := functionframeworks.RegisterCloudEventFunction(ctx, fnCloudEvent); err != nil {
+            return fmt.Errorf("Function failed to register: %v\n", err)
+        }
+    } else if fnOpenFunction, ok := fn.(func(*ofctx.OpenFunctionContext, []byte) ofctx.RetValue); ok {
+        if err := functionframeworks.RegisterOpenFunction(ctx, fnOpenFunction); err != nil {
+            return fmt.Errorf("Function failed to register: %v\n", err)
+        }
+    } else {
+        err := errors.New("unrecognized function")
+        return fmt.Errorf("Function failed to register: %v\n", err)
+    }
+    return nil
+}
+
+func main() {
+    if err := register(userfunction.HelloWorld); err != nil {
+        log.Fatalf("Failed to register: %v\n", err)
+    }
+
+    if err := functionframeworks.Start(); err != nil {
+        log.Fatalf("Failed to start: %v\n", err)
+    }
+}
+```
+
+Before the application is started, the function is registered as an HTTP function, a CloudEvent function, or an OpenFunction function. When the registration is complete, `functionframeworks.Start` is invoked to start the application.
```

Comment:
complete => completed?

Reply:
They mean the same thing, but I'll update this word per your comment to avoid possible ambiguity.

Reply:
Updated.